snake-2/snake.js

```diff
@@ -0,0 +1,307 @@
+
+window.onload = () => createSnakeEngine(
+    "snakeCanvas", 
+    "foodCounter",
+    20,
+    200,
+    30,
+    getSnakeCoordinates);
+
+
+var snake = {
+    x: 0,
+    y: 0,
+    direction: "down"
+}
+
+
+function getSnakeCoordinates(mapSize, foodCoordinates) {
+   
+    let closestFoodIndices = findClosestFood(snake, foodCoordinates);
+    newHeading(snake, closestFoodIndices);
+
+    if(snake.direction == "down") {
+        var newY = snake.y + 1;
+
+        if(newY >= mapSize) {
+            snake.direction = "right";
+        } else {
+            snake.y = newY;
+        }
+    }
+
+    if(snake.direction == "right") {
+        var newX = snake.x + 1;
+
+        if(newX >= mapSize) {
+            snake.direction = "up";
+        } else {
+            snake.x = newX;
+        }
+    }
+
+    if(snake.direction == "up") {
+        var newY = snake.y - 1;
+
+        if(newY < 0) {
+            snake.direction = "left";
+        } else {
+            snake.y = newY;
+        }
+    }
+
+    if(snake.direction == "left") {
+        var newX = snake.x - 1;
+
+        if(newX < 0) {
+            snake.direction = "down";
+        } else {
+            snake.x = newX;
+        }
+    }
+
+    return [{
+        x: snake.x,
+        y: snake.y
+    }];
+
+}
+
+function findClosestFood(snake, foodCoordinates) {
+
+    let squaredDiff = []; //tablica kwadratow odleglosci miedzy niebieska, a czerwona kulka
+
+    for (let index = 0; index < foodCoordinates.length; index++) {
+        squaredDiff[index] = (snake.x - foodCoordinates[index].x)**2 + (snake.y - foodCoordinates[index].y)**2;
+    }
+
+    let minSquaredDiff = Math.min.apply(Math, squaredDiff);
+    let indexMinSquaredDiff = squaredDiff.indexOf(minSquaredDiff);
+
+    return [foodCoordinates[indexMinSquaredDiff].x, foodCoordinates[indexMinSquaredDiff].y];
+
+}
+
+function newHeading(snake, closestFood) {
```

Comment:
Dodatkowo, warto by zmodyfikować tą funkcję tak, by zamiast zmieniać snake'a którego przyjmuje w argumencie, po prostu zwracała nowy kierunek, tzn np.

return 'down';

Najlepsza funkcja to taka, która nie modyfikuje swoich wejść a tylko zwraca jakieś wyjście

Reply:
poprawione